File: packages/ts/src/discover.test.ts
import { mkdirSync, realpathSync } from 'node:fs'
import { join } from 'node:path'
import { afterAll, describe, expect, it } from 'vitest'

import { listConfigFiles, resolveProjectRoots, resolveUserRoot } from './discover.js'
import { ConfigError } from './errors.js'
import { cleanupTrees, makeTree, warningCollector } from './test-utils.js'

afterAll(cleanupTrees)

describe('resolveProjectRoots', () => {
  it('finds a config directory two levels up', () => {
    const tree = makeTree({
      '.git': '',
      '.config/mytool/config.toml': 'a = 1\n',
      'a/b/.gitkeep': '',
    })
    expect(resolveProjectRoots(join(tree, 'a/b'), 'mytool')).toEqual([
      join(realpathSync(tree), '.config/mytool'),
    ])
  })

  it('returns roots farthest ancestor first', () => {
    const tree = makeTree({
      '.git': '',
      '.config/mytool/config.toml': 'a = 1\n',
      'pkg/.config/mytool/config.toml': 'a = 2\n',
      'pkg/src/.gitkeep': '',
    })
    const real = realpathSync(tree)
    // Order is the contract: the nearest root is last, and last is what wins.
    expect(resolveProjectRoots(join(tree, 'pkg/src'), 'mytool')).toEqual([
      join(real, '.config/mytool'),
      join(real, 'pkg/.config/mytool'),
    ])
  })

  it('stops at a .git directory between cwd and the config', () => {
    const tree = makeTree({
      '.config/mytool/config.toml': 'a = 1\n',
      'repo/.git/HEAD': 'ref: refs/heads/main\n',
      'repo/pkg/.gitkeep': '',
    })
    expect(resolveProjectRoots(join(tree, 'repo/pkg'), 'mytool')).toEqual([])
  })

  it('stops at a .git *file*, the form git uses for a worktree or submodule', () => {
    const tree = makeTree({
      '.config/mytool/config.toml': 'a = 1\n',
      'repo/.git': 'gitdir: /elsewhere/.git/worktrees/w\n',
      'repo/pkg/.gitkeep': '',
    })
    expect(resolveProjectRoots(join(tree, 'repo/pkg'), 'mytool')).toEqual([])
  })

  it('checks the stopDir itself before ending the walk', () => {
    const tree = makeTree({
      '.git': '',
      'pkg/.config/mytool/config.toml': 'a = 1\n',
      'pkg/src/.gitkeep': '',
    })
    const real = realpathSync(tree)
    // stopDir is inclusive: the directory holding the config is searched, then the walk ends.
    expect(
      resolveProjectRoots(join(tree, 'pkg/src'), 'mytool', { stopDir: join(tree, 'pkg') }),
    ).toEqual([join(real, 'pkg/.config/mytool')])
  })

  it('stops at the injected home directory, inclusive', () => {
    const tree = makeTree({
      '.git': '',
      'home/.config/mytool/config.toml': 'a = 1\n',
      'home/work/.gitkeep': '',
    })
    const real = realpathSync(tree)
    expect(
      resolveProjectRoots(join(tree, 'home/work'), 'mytool', { home: join(tree, 'home') }),
    ).toEqual([join(real, 'home/.config/mytool')])
  })

  it('raises rather than spinning when the walk goes pathologically deep', () => {
    const tree = makeTree({})
    const deep = join(tree, Array.from({ length: 70 }, (_, i) => `d${i}`).join('/'))
    mkdirSync(deep, { recursive: true })
    expect(() => resolveProjectRoots(deep, 'mytool')).toThrow(/exceeded 64 directories/)
  })
})

describe('resolveUserRoot', () => {
  it('prefers an absolute XDG_CONFIG_HOME over the home directory', () => {
    const tree = makeTree({
      'xdg/mytool/config.toml': 'a = 1\n',
      'home/.config/mytool/config.toml': 'a = 2\n',
    })
    expect(
      resolveUserRoot('mytool', {
        home: join(tree, 'home'),
        env: { XDG_CONFIG_HOME: join(tree, 'xdg') },
      }),
    ).toBe(join(tree, 'xdg/mytool'))
  })

  it('ignores a relative XDG_CONFIG_HOME, warns, and falls back to home', () => {
    const tree = makeTree({ 'home/.config/mytool/config.toml': 'a = 1\n' })
    const { warnings, onWarning } = warningCollector()
    const root = resolveUserRoot('mytool', {
      home: join(tree, 'home'),
      env: { XDG_CONFIG_HOME: '../cfg' },
      onWarning,
    })
    expect(root).toBe(join(tree, 'home/.config/mytool'))
    expect(warnings).toHaveLength(1)
    expect(warnings[0]).toMatch(/XDG_CONFIG_HOME.*absolute/)
  })

  it('ignores an empty XDG_CONFIG_HOME and warns', () => {
    const tree = makeTree({ 'home/.config/mytool/config.toml': 'a = 1\n' })
    const { warnings, onWarning } = warningCollector()
    expect(
      resolveUserRoot('mytool', { home: join(tree, 'home'), env: { XDG_CONFIG_HOME: '' }, onWarning }),
    ).toBe(join(tree, 'home/.config/mytool'))
    expect(warnings).toHaveLength(1)
  })

  it('uses home/.config when XDG_CONFIG_HOME is unset', () => {
    const tree = makeTree({ 'home/.config/mytool/config.toml': 'a = 1\n' })
    const { warnings, onWarning } = warningCollector()
    expect(resolveUserRoot('mytool', { home: join(tree, 'home'), env: {}, onWarning })).toBe(
      join(tree, 'home/.config/mytool'),
    )
    expect(warnings).toEqual([])
  })

  it('returns null when the resolved directory does not exist', () => {
    const tree = makeTree({ 'home/.gitkeep': '' })
    expect(resolveUserRoot('mytool', { home: join(tree, 'home'), env: {} })).toBeNull()
  })
})

describe('listConfigFiles', () => {
  it('returns recognized files in the SPEC §2.5 order', () => {
    const tree = makeTree({
      'c/.env': 'A=1\n',
      'c/config.ini': '[s]\nk = 1\n',
      'c/config.json': '{}\n',
      'c/config.jsonc': '{}\n',
      'c/config.toml': 'a = 1\n',
      'c/config.yaml': 'a: 1\n',
      'c/unrecognized.txt': 'ignored\n',
    })
    expect(listConfigFiles(join(tree, 'c')).map((f) => [f.path.split('/').pop(), f.format])).toEqual([
      ['config.toml', 'toml'],
      ['config.yaml', 'yaml'],
      ['config.json', 'json'],
      ['config.jsonc', 'jsonc'],
      ['config.ini', 'ini'],
      ['.env', 'dotenv'],
    ])
  })

  it('raises duplicate-format for config.yaml beside config.yml', () => {
    const tree = makeTree({ 'c/config.yaml': 'a: 1\n', 'c/config.yml': 'a: 2\n' })
    try {
      listConfigFiles(join(tree, 'c'))
      expect.unreachable('expected a duplicate-format error')
    } catch (error) {
      expect(error).toBeInstanceOf(ConfigError)
      expect((error as ConfigError).kind).toBe('duplicate-format')
      expect((error as ConfigError).path).toBe(join(tree, 'c'))
    }
  })

  it('loads a profile file immediately after its base file', () => {
    const tree = makeTree({ 'c/config.toml': 'a = 1\n', 'c/config.prod.toml': 'a = 2\n' })
    expect(listConfigFiles(join(tree, 'c'), 'prod').map((f) => f.path.split('/').pop())).toEqual([
      'config.toml',
      'config.prod.toml',
    ])
  })
})

File: packages/golang/discover_test.go
package configdiscovery

import (
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

// tree materializes {relative path: contents} under t.TempDir() and returns the root, resolved
// so comparisons against the walk's output line up on macOS where /var is a symlink.
//
// Every test here builds its own tree and injects cwd and home. Nothing reads the real working
// directory or the real home: a suite whose results depend on the machine is not a suite.
func tree(t *testing.T, files map[string]string) string {
	t.Helper()
	root := t.TempDir()
	for relative, contents := range files {
		target := filepath.Join(root, relative)
		if strings.HasSuffix(relative, "/") {
			if err := os.MkdirAll(target, 0o755); err != nil {
				t.Fatal(err)
			}
			continue
		}
		if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
			t.Fatal(err)
		}
		if err := os.WriteFile(target, []byte(contents), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	resolved, err := filepath.EvalSymlinks(root)
	if err != nil {
		t.Fatal(err)
	}
	return resolved
}

func TestResolveProjectRootsFindsAConfigTwoLevelsUp(t *testing.T) {
	root := tree(t, map[string]string{
		".git":                       "",
		".config/mytool/config.toml": "a = 1\n",
		"a/b/":                       "",
	})
	got, err := resolveProjectRoots(filepath.Join(root, "a/b"), "mytool", walkOptions{})
	if err != nil {
		t.Fatal(err)
	}
	want := []string{filepath.Join(root, ".config/mytool")}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v, want %v", got, want)
	}
}

func TestResolveProjectRootsOrdersFarthestAncestorFirst(t *testing.T) {
	root := tree(t, map[string]string{
		".git":                           "",
		".config/mytool/config.toml":     "a = 1\n",
		"pkg/.config/mytool/config.toml": "a = 2\n",
		"pkg/src/":                       "",
	})
	got, err := resolveProjectRoots(filepath.Join(root, "pkg/src"), "mytool", walkOptions{})
	if err != nil {
		t.Fatal(err)
	}
	// Order is the contract: the nearest root is last, and last is what wins.
	want := []string{
		filepath.Join(root, ".config/mytool"),
		filepath.Join(root, "pkg/.config/mytool"),
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v, want %v", got, want)
	}
}

func TestResolveProjectRootsStopsAtAGitDirectory(t *testing.T) {
	root := tree(t, map[string]string{
		".config/mytool/config.toml": "a = 1\n",
		"repo/.git/HEAD":             "ref: refs/heads/main\n",
		"repo/pkg/":                  "",
	})
	got, err := resolveProjectRoots(filepath.Join(root, "repo/pkg"), "mytool", walkOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Fatalf("expected the walk to stop before the config, got %v", got)
	}
}

func TestResolveProjectRootsStopsAtAGitFile(t *testing.T) {
	// The form git writes for a linked worktree or a submodule. Those are repositories too, and
	// a walk that only tests for a directory climbs straight past them.
	root := tree(t, map[string]string{
		".config/mytool/config.toml": "a = 1\n",
		"repo/.git":                  "gitdir: /elsewhere/.git/worktrees/w\n",
		"repo/pkg/":                  "",
	})
	got, err := resolveProjectRoots(filepath.Join(root, "repo/pkg"), "mytool", walkOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Fatalf("expected the walk to stop before the config, got %v", got)
	}
}

func TestResolveProjectRootsStopDirIsInclusive(t *testing.T) {
	root := tree(t, map[string]string{
		".git":                           "",
		"pkg/.config/mytool/config.toml": "a = 1\n",
		"pkg/src/":                       "",
	})
	got, err := resolveProjectRoots(
		filepath.Join(root, "pkg/src"), "mytool", walkOptions{stopDir: filepath.Join(root, "pkg")},
	)
	if err != nil {
		t.Fatal(err)
	}
	want := []string{filepath.Join(root, "pkg/.config/mytool")}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v, want %v", got, want)
	}
}

func TestResolveProjectRootsHomeIsInclusive(t *testing.T) {
	root := tree(t, map[string]string{
		".git":                            "",
		"home/.config/mytool/config.toml": "a = 1\n",
		"home/work/":                      "",
	})
	got, err := resolveProjectRoots(
		filepath.Join(root, "home/work"), "mytool", walkOptions{home: filepath.Join(root, "home")},
	)
	if err != nil {
		t.Fatal(err)
	}
	want := []string{filepath.Join(root, "home/.config/mytool")}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v, want %v", got, want)
	}
}

func TestResolveProjectRootsDepthCap(t *testing.T) {
	root := t.TempDir()
	deep := root
	for i := 0; i < 70; i++ {
		deep = filepath.Join(deep, "d")
	}
	if err := os.MkdirAll(deep, 0o755); err != nil {
		t.Fatal(err)
	}
	_, err := resolveProjectRoots(deep, "mytool", walkOptions{})
	if err == nil || !strings.Contains(err.Error(), "exceeded 64 directories") {
		t.Fatalf("expected the depth cap to fire, got %v", err)
	}
}

func TestResolveUserRoot(t *testing.T) {
	root := tree(t, map[string]string{
		"xdg/mytool/config.toml":          "a = 1\n",
		"home/.config/mytool/config.toml": "a = 2\n",
		"emptyhome/":                      "",
	})

	cases := []struct {
		name        string
		home        string
		env         map[string]string
		want        string
		wantWarning bool
	}{
		{
			name: "an absolute XDG_CONFIG_HOME wins",
			home: filepath.Join(root, "home"),
			env:  map[string]string{"XDG_CONFIG_HOME": filepath.Join(root, "xdg")},
			want: filepath.Join(root, "xdg/mytool"),
		},
		{
			name:        "a relative XDG_CONFIG_HOME is ignored and warns",
			home:        filepath.Join(root, "home"),
			env:         map[string]string{"XDG_CONFIG_HOME": "../cfg"},
			want:        filepath.Join(root, "home/.config/mytool"),
			wantWarning: true,
		},
		{
			name:        "an empty XDG_CONFIG_HOME is ignored and warns",
			home:        filepath.Join(root, "home"),
			env:         map[string]string{"XDG_CONFIG_HOME": ""},
			want:        filepath.Join(root, "home/.config/mytool"),
			wantWarning: true,
		},
		{
			name: "unset XDG_CONFIG_HOME uses home/.config",
			home: filepath.Join(root, "home"),
			env:  map[string]string{},
			want: filepath.Join(root, "home/.config/mytool"),
		},
		{
			name: "a missing directory is no user root at all",
			home: filepath.Join(root, "emptyhome"),
			env:  map[string]string{},
			want: "",
		},
	}

	for _, testCase := range cases {
		t.Run(testCase.name, func(t *testing.T) {
			var warnings []string
			got := resolveUserRoot("mytool", testCase.home, testCase.env, func(message string) {
				warnings = append(warnings, message)
			})
			if got != testCase.want {
				t.Fatalf("got %q, want %q", got, testCase.want)
			}
			if testCase.wantWarning && len(warnings) != 1 {
				t.Fatalf("expected exactly one warning, got %v", warnings)
			}
			if !testCase.wantWarning && len(warnings) != 0 {
				t.Fatalf("expected no warning, got %v", warnings)
			}
		})
	}
}
